lms/djangoapps/courseware/field_overrides.py

```diff
@@ -171,83 +256,54 @@ def default(self, block, name):
-    must provide.  In general, providers should derive from this class, but
-    it's not strictly necessary as long as they correctly implement this
-    interface.
+        Arguments:
+            block: An XBlock
+            field_data: An instance of FieldData to be wrapped
+        """
+        if cls.provider_classes is None:
+            cls.provider_classes = [
+                resolve_dotted(name) for name in settings.MODULESTORE_FIELD_OVERRIDE_PROVIDERS
+            ]
 
-    A `FieldOverrideProvider` implementation is only responsible for looking up
-    field overrides. To set overrides, there will be a domain specific API for
-    the concrete override implementation being used.
-    """
-    __metaclass__ = ABCMeta
+        enabled_providers = cls._providers_for_block(block)
+        if enabled_providers:
+            return cls(field_data, enabled_providers)
 
-    def __init__(self, user):
-        self.user = user
+        return field_data
 
-    @abstractmethod
-    def get(self, block, name, default):  # pragma no cover
-        """
-        Look for an override value for the field named `name` in `block`.
-        Returns the overridden value or `default` if no override is found.
+    @classmethod
+    def _providers_for_block(cls, block):
         """
-        raise NotImplementedError
+        Computes a list of enabled providers based on the given XBlock.
+        The result is cached per request to avoid the overhead incurred
+        by filtering override providers hundreds of times.
 
-    @abstractmethod
-    def enabled_for(self, course):  # pragma no cover
+        Arguments:
+            block: An XBlock
         """
-        Return True if this provider should be enabled for a given course,
-        and False otherwise.
-
-        Concrete implementations are responsible for implementing this method.
+        course_id = unicode(block.location.course_key)
+        cache_key = ENABLED_MODULESTORE_OVERRIDE_PROVIDERS_KEY.format(course_id=course_id)
 
-        Arguments:
-          course (CourseModule or None)
+        request_cache = RequestCache.get_request_cache()
+        enabled_providers = request_cache.data.get(cache_key)
 
-        Returns:
-          bool
-        """
-        return False
+        if enabled_providers is None:
+            enabled_providers = [
+                provider_class for provider_class in cls.provider_classes if provider_class.enabled_for(block)
+            ]
+            request_cache.data[cache_key] = enabled_providers
 
+        return enabled_providers
 
-def _lineage(block):
-    """
-    Returns an iterator over all ancestors of the given block, starting with
-    its immediate parent and ending at the root of the block tree.
-    """
-    parent = block.get_parent()
-    while parent:
-        yield parent
-        parent = parent.get_parent()
+    def __init__(self, fallback, providers):
+        super(OverrideModulestoreFieldData, self).__init__(None, fallback, providers)
```

Comment:
@cpennington I've left this `__init__` in place because it explicitly provides `None` as the value of `super`'s `user` argument.